### ISSUE TITLE
Inconsistent dropout implementation across models

### ISSUE THREAD

**Issue by rvashurin** [huggingface/transformers#26896]
### Feature request

Dropout implementations vary across different models - for example, T5 implements dropout as a layer https://github.com/huggingface/transformers/blob/e893b1efbbce80b7eaaf24f9e0134450820782b5/src/transformers/models/t5/modeling_t5.py#L285 which is visible in the model printout, while Bart uses functional form of dropout https://github.com/huggingface/transformers/blob/e893b1efbbce80b7eaaf24f9e0134450820782b5/src/transformers/models/bart/modeling_bart.py#L434 which is essentially hidden from the user. It would be useful if either all models had the same (preferably, layer-like) implementation of dropout, or models had flag indicating how dropout is implemented in a model.

### Motivation

I am currently working with some Monte-Carlo sampling from parameter distributions of models, which requires a re-implementation of dropout to essentially repeat masks across batch and beam dimensions, instead of sampling fully random mask for all tensor dimensions. This is a trivial task if the model has separate layer for dropout operation, but not if it uses functional form. The latter case requires one to patch `torch.nn.functional.dropout` during runtime which is as dirty a hack as they come. I understand that this is a very niche problem, yet this would be beneficial from implementation consistency perspective if nothing else.

### Your contribution

I could work on solution if necessary and if HF staff signals that this can possibly be merged.

**Comment by ArthurZucker** [comment]
I don't really see a reason why not to standardize the dropout to improve consistency ad readability, but that would probably require us to change quite a lot of models (380~ lines use `nn.functional.dropout` and 758 use `self.dropout = nn.Dropout` )  
Wondering rather if there are any specific known reason to use one rather than the other in the first place. 

FYI @LysandreJik

**Comment by LysandreJik** [comment]
Not an issue for me. We did something similar for activation functions ~3 years ago. I'd just make sure it's backwards compatible and that it's the most versatile (in this case, the one that can be patched)

**Comment by rvashurin** [comment]
I will try to assess what the best approach is and conjure up some draft PR soon-ish.
